### PR TITLE
server: add terminate_handler to log on uncaught exception

### LINKED ISSUE #1378: Add support for printing stack traces when exception is unhandled
This isn't a feature that can be provided in a portable fashion, but it very useful where the compiler/runtime don't unwind the stack before calling the std::terminate_handler when dying.

### PR DESCRIPTION
**Pull request by gumpt**
Adds a `terminate_handler` to log a backtrace if possible.

This keeps using the third-party dependency "backward" -- it's used already, so I thought it best to keep using it.

*Risk Level*: Low

*Testing*: Wrote a death test which calls `std::terminate`.  Added (then removed) an uncaught exception, saw this show up in the trace.

*Docs Changes*: N/A

*Release Notes*: N/A

Fixes #1378

### CODE REVIEW

**Comment by mattklein123** [comment]
@gumpt thanks for working on this. We definitely need this to be guarded by `--define signal_trace=disabled` as in production environments some folks are not going to want this behavior (including Lyft). Can we guard this in similar ways as to the existing backtrace functionality?

**Comment by gumpt** [comment]
That should do it, and CI seems happy.

**Comment by mattklein123** [comment]
@dnoe can you review? @gumpt it would be nice to have a death test for this. I would take a look at similar tests that @dnoe wrote already for the backtrace code.

**Comment by dnoe** [comment]
Sorry, this dropped off my radar and got buried in GH emails when I was on vacation last week. Will review by this afternoon.

File: test/exe/terminate_handler_test.cc [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#include "exe/terminate_handler.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+TEST(TerminateHandler, HandlerInstalledTest) {
+  TerminateHandler handler;
+  EXPECT_DEATH([]() -> void { std::terminate(); }(),
```
Comment (mattklein123) [inline test/exe/terminate_handler_test.cc]: note: please merge master, you need to use a different macro now for death tests (search code, don't remember off the top of my head)

Comment (dnoe) [reply test/exe/terminate_handler_test.cc]: I think you want EXPECT_DEATH_LOG_TO_STDERR

File: test/exe/terminate_handler_test.cc [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+               ".*Envoy::TerminateHandler::logOnTerminate().*|backtrace\\.h");
+}
```
Comment (dnoe) [inline test/exe/terminate_handler_test.cc]: I was going to ask if you could write a test that actually throws, but I tried that and gmock catches it first and helpfully indicates the test failed due to unhandled exception :) Oh well...

File: source/exe/terminate_handler.cc [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <cstdlib>
+
+#include "server/backtrace.h"
+
+namespace Envoy {
+
+std::terminate_handler TerminateHandler::logOnTerminate() const {
+  return std::set_terminate([]() {
+    BACKTRACE_LOG();
```
Comment (dnoe) [inline source/exe/terminate_handler.cc]: Once you look at the addresses it's obvious that the backtrace and abort are due to terminate(), but that takes a bit of extra effort. How about an extra log statement here indicating that an uncaught exception occurred? That'll make it very obvious. Unfortunately the C++ standard doesn't allow the terminate handler to figure out useful things like the type of the exception or other exception details...

File: test/exe/terminate_handler_test.cc (hunk 2 above)
Comment (gumpt) [reply test/exe/terminate_handler_test.cc]: I tried to do this originally too. :-(

**Comment by gumpt** [comment]
Think that addresses all of the new feedback.  Thanks for not letting this die on the vine!

**Pull request merged**